src/protocol/modulus_conversion/convert_shares.rs

```diff
@@ -48,69 +59,164 @@ impl crate::protocol::step::Step for Step {}
+    /// without truncation or if the bit index is out of range for `B`.
+    #[must_use]
+    pub fn new(helper_role: Role, left: bool, right: bool) -> Self {
+        let left = F::try_from(u128::from(left)).unwrap();
+        let right = F::try_from(u128::from(right)).unwrap();
+        Self(match helper_role {
+            Role::H1 => [
+                Replicated::new(left, F::ZERO),
+                Replicated::new(F::ZERO, right),
+                Replicated::new(F::ZERO, F::ZERO),
+            ],
+            Role::H2 => [
+                Replicated::new(F::ZERO, F::ZERO),
+                Replicated::new(left, F::ZERO),
+                Replicated::new(F::ZERO, right),
+            ],
+            Role::H3 => [
+                Replicated::new(F::ZERO, right),
+                Replicated::new(F::ZERO, F::ZERO),
+                Replicated::new(left, F::ZERO),
+            ],
+        })
+    }
 }
 
-#[must_use]
-pub fn convert_all_bits_local<F: Field, B: GaloisField>(
-    helper_role: Role,
-    input: impl Iterator<Item = Replicated<B>>,
-) -> Vec<BitDecomposed<BitConversionTriple<Replicated<F>>>> {
-    input
-        .map(move |record| {
-            BitDecomposed::decompose(B::BITS, |bit_index| {
-                convert_bit_local::<F, B>(helper_role, bit_index, &record)
-            })
-        })
-        .collect::<Vec<_>>()
+pub trait ToBitConversionTriples {
+    /// Get the maximum number of bits that can be produced for this type.
+    ///
+    /// Note that this should be an associated constant, but one of the implementations would then need
+    /// const generics to be more fully available in the language, so this is a method instead.  For now.
+
+    fn bits(&self) -> u32;
+    /// Produce a `BitConversionTriple` for the given role and bit index.
+    fn triple<F: PrimeField>(&self, role: Role, i: u32) -> BitConversionTriple<Replicated<F>>;
+
+    fn triple_range<F, I>(
+        &self,
+        role: Role,
+        indices: I,
+    ) -> BitDecomposed<BitConversionTriple<Replicated<F>>>
+    where
+        F: PrimeField,
+        I: IntoIterator<Item = u32>,
+    {
+        BitDecomposed::new(indices.into_iter().map(|i| self.triple(role, i)))
+    }
+}
+
+impl<B: GaloisField> ToBitConversionTriples for Replicated<B> {
+    fn bits(&self) -> u32 {
+        B::BITS
+    }
+
+    fn triple<F: PrimeField>(&self, role: Role, i: u32) -> BitConversionTriple<Replicated<F>> {
+        BitConversionTriple::new(role, self.left()[i], self.right()[i])
+    }
+}
+
+impl ToBitConversionTriples for BitDecomposed<Replicated<Gf2>> {
+    fn bits(&self) -> u32 {
+        u32::try_from(self.len()).unwrap()
+    }
+
+    fn triple<F: PrimeField>(&self, role: Role, i: u32) -> BitConversionTriple<Replicated<F>> {
+        const BIT0: u32 = 0;
+        let i = usize::try_from(i).unwrap();
+        BitConversionTriple::new(role, self[i].left()[BIT0], self[i].right()[BIT0])
+    }
+}
+
+#[pin_project]
```

Comment:
See https://doc.rust-lang.org/std/pin/index.html#projections-and-structural-pinning

When you have `Pin<&mut X>` you can't really do anything with the contents of the struct X because `Pin` stops you from accessing the pointer for `X` directly.  For all the fields in `X`, you have two choices:

1. Turn them into their own `Pin<&mut Y>`.  This says that it is very important that `X` stay in the same place because `Y` also needs to stay in the same place.  For this, which is a wrapper around a stream, that's why we use `#[pin]` for the stream we're wrapping.  Also, most of the APIs take `Pin<&mut S> where S: Stream`.
2. Provide an accessor for `&mut Z`.  For fields like the `role` and `bits`, we don't need to worry about that moving about, so we don't bother keeping that pinned down.

The `pin_project` crate provides a wrapper that allows us to do all of that automatically, without `unsafe{}` blocks.  This mostly just makes it all safe.